include/log.h

```diff
@@ -25,68 +21,41 @@
 #define CELERITY_ERROR(...) CELERITY_LOG(::celerity::detail::log_level::err, __VA_ARGS__)
 #define CELERITY_CRITICAL(...) CELERITY_LOG(::celerity::detail::log_level::critical, __VA_ARGS__)
 
-namespace celerity {
-namespace detail {
+namespace celerity::detail {
 
-	using log_level = spdlog::level::level_enum;
+using log_level = spdlog::level::level_enum;
+using spdlog::should_log;
 
-	template <typename... Es>
-	struct log_map {
-		const std::tuple<Es...>& entries;
-		log_map(const std::tuple<Es...>& entries) : entries(entries) {}
-	};
-
-	struct log_context {
-		std::string value;
-		log_context() = default;
-		template <typename... Es>
-		explicit log_context(const std::tuple<Es...>& entries) {
-			static_assert(sizeof...(Es) % 2 == 0, "log_context requires key/value pairs");
-			value = fmt::format("[{}] ", log_map{entries});
-		}
-	};
-
-	inline const std::string null_log_ctx;
-	inline thread_local const std::string* active_log_ctx = &null_log_ctx;
-
-	struct log_ctx_setter {
-		log_ctx_setter(log_context& ctx) { celerity::detail::active_log_ctx = &ctx.value; }
-		~log_ctx_setter() { celerity::detail::active_log_ctx = &celerity::detail::null_log_ctx; }
-	};
+struct log_context {
+	std::string repr;
 
-#define CELERITY_DETAIL_LOG_SET_SCOPED_CTX(ctx)                                                                                                                \
-	log_ctx_setter _set_log_ctx_##__COUNTER__ { ctx }
+	log_context() = default;
 
-	template <typename Tuple, typename Callback>
-	constexpr void tuple_for_each_pair_impl(const Tuple&, Callback&&, std::index_sequence<>) {}
-
-	template <typename Tuple, size_t I1, size_t I2, size_t... Is, typename Callback>
-	constexpr void tuple_for_each_pair_impl(const Tuple& tuple, const Callback& cb, std::index_sequence<I1, I2, Is...>) {
-		cb(std::get<I1>(tuple), std::get<I2>(tuple));
-		tuple_for_each_pair_impl(tuple, cb, std::index_sequence<Is...>{});
-	}
-
-	template <typename Tuple, typename Callback>
-	constexpr void tuple_for_each_pair(const Tuple& tuple, const Callback& cb) {
-		static_assert(std::tuple_size_v<Tuple> % 2 == 0, "an even number of entries is required");
-		tuple_for_each_pair_impl(tuple, cb, std::make_index_sequence<std::tuple_size_v<Tuple>>{});
+	template <typename... Es>
+	explicit log_context(const std::tuple<Es...>& entries) {
+		static_assert(sizeof...(Es) % 2 == 0, "log_context requires key/value pairs");
+		int i = 0;
+		celerity::detail::utils::tuple_for_each_pair(entries, [&](auto& a, auto& b) {
+			if(i++ > 0) { fmt::format_to(std::back_inserter(repr), ", "); }
+			fmt::format_to(std::back_inserter(repr), "{}={}", a, b);
+		});
```

Comment:
```suggestion
		std::string keys_and_values;
		celerity::detail::utils::tuple_for_each_pair(entries, [&](auto& a, auto& b) {
			if(i++ > 0) { fmt::format_to(std::back_inserter(keys_and_values), ", "); }
			fmt::format_to(std::back_inserter(keys_and_values), "{}={}", a, b);
		});
		if(i > 0) { repr = fmt::format("[{}] ", keys_and_values); }
```

Square brackets and final space went missing!

Reply:
Done!